File: bfsdfs/islandproblems/noofisland/NoOfIsland.py
from collections import deque

class NoOfIsland:
    def numIslands(self, grid):
        if not grid or len(grid) == 0:
            return 0

        num_islands = 0
        num_rows = len(grid)
        num_cols = len(grid[0])

        # Directions for moving up, down, left, right
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        def bfs(start_row, start_col):
            queue = deque([(start_row, start_col)])
            grid[start_row][start_col] = '0'  # Mark as visited

            while queue:
                current_row, current_col = queue.popleft()

                for direction in directions:
                    new_row = current_row + direction[0]
                    new_col = current_col + direction[1]

                    # Check if the new position is valid and is land
                    if 0 <= new_row < num_rows and 0 <= new_col < num_cols and grid[new_row][new_col] == '1':
                        grid[new_row][new_col] = '0'  # Mark as visited
                        queue.append((new_row, new_col))

        # Iterate through each cell in the grid
        for row in range(num_rows):
            for col in range(num_cols):
                if grid[row][col] == '1':  # Found an island
                    num_islands += 1
                    bfs(row, col)  # Start BFS to mark the entire island

        return num_islands


# Example usage
if __name__ == "__main__":
    grid = [
        ["1","1","0","0","0"],
        ["1","1","0","0","0"],
        ["0","0","1","0","0"],
        ["0","0","0","1","1"]
    ]
    solution = NoOfIsland()
    result = solution.numIslands(grid)
    print("Number of islands:", result)

File: bfsdfs/islandproblems/noofisland/NoOfIsland.java
package islandproblems.noofisland;
import java.util.LinkedList;
import java.util.Queue;

public class NoOfIsland {
    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0) {
            return 0;
        }

        int numIslands = 0;
        int rows = grid.length;
        int cols = grid[0].length;

        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                // If we find a '1', we have found an island
                if (grid[row][col] == '1') {
                    numIslands++;
                    // Perform BFS to mark all parts of this island as visited
                    bfs(grid, row, col);
                }
            }
        }

        return numIslands;
    }

    private void bfs(char[][] grid, int startRow, int startCol) {
        // Directions for moving up, down, left, right
        int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        // Create a queue for BFS
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{startRow, startCol});

        // Mark the starting point as visited
        grid[startRow][startCol] = '0'; // Change '1' to '0' to mark it as visited

        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int currentRow = current[0];
            int currentCol = current[1];

            // Check all four directions
            for (int[] direction : directions) {
                int newRow = currentRow + direction[0];
                int newCol = currentCol + direction[1];

                // If the new position is valid and is part of the island
                if (newRow >= 0 && newRow < grid.length && newCol >= 0 && newCol < grid[0].length && grid[newRow][newCol] == '1') {
                    queue.offer(new int[]{newRow, newCol});
                    grid[newRow][newCol] = '0'; // Mark as visited
                }
            }
        }
    }

    public static void main(String[] args) {
        NoOfIsland solution = new NoOfIsland();
        char[][] grid = {
                {'1', '1', '1', '1', '0'},
                {'1', '1', '0', '1', '0'},
                {'1', '1', '0', '0', '0'},
                {'0', '0', '0', '0', '0'}
        };

        char[][] grid1 = {
                {'1', '1', '0', '0', '0'},
                {'1', '1', '0', '0', '0'},
                {'0', '0', '1', '0', '0'},
                {'0', '0', '0', '1', '1'}
        };
        int result = solution.numIslands(grid);
        System.out.println("Number of islands for grid: " + result);
        System.out.println("Number of islands for grid1: " + solution.numIslands(grid1));
    }
}
